src/main/java/commission/service/PaymentCalculatorService.java

```diff
@@ -0,0 +1,61 @@
+package commission.service;
+
+import commission.dao.CrudDao;
+import commission.entity.Commission;
+import commission.entity.Sale;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+public class PaymentCalculatorService {
+
+    private  final CrudDao<Sale> saleDao;
+    private final CrudDao<Commission> commissionDao;
+
+    public PaymentCalculatorService(CrudDao<Sale> saleDao, CrudDao<Commission> commissionDao) {
+        this.saleDao = saleDao;
+        this.commissionDao = commissionDao;
+    }
+
+    public void completeCommission(Commission commission){
+        Optional<Sale> sale = this.saleDao.findById(commission.salesId());
+        Commission completeCommission = new Commission(
+                commission.id(),
+                commission.commissionType(),
+                calculateCommissionPayment(commission),
+                commission.salesId());
+
+       this.commissionDao.insert(completeCommission);
+    }
+
+    public double calculateCommissionPayment(Commission commission){
+        Optional<Sale> sale = this.saleDao.findById(commission.salesId());
+        var price = sale.get().price();
```

Comment:
After writing this comment I realised that there is referential integrity at the database level for this and the Sale record should always exist for the commission... So you can IGNORE this comment but I will leave it here simply so you can read and learn

Simply calling `.get()` on an `Optional` runs the risk of `NullPointerException` since it could return null when nothing is there. There are a few ways you can handle the `Optional` like adding an if statement like you have in other services. The two options I prefer using are the `.orElse` and `.orElseThrow` methods. `.ofElse` returns a default value if the Optional is empty or contains null so you will always have a value to work with. `.orElseThrow` throws an exception if the Option is empty, and in this case (in my opinion at least) we want to throw an exception because the sale was not found.
```
var sale = this.saleDao
    .findById(commission.salesId())
    .orElseThrow(() -> new ErrorResponseException(HttpStatus.NOT_FOUND));
```

Reply:
got it. Also learned that i can do this 
`.orElseThrow(() -> new ErrorResponseException(HttpStatus.NOT_FOUND));` 

Would this way be better to `findById` or just return without assigning to a variable

```
public Optional<T> findByIdO(long id) {
        var temp = someDao
                .findById(id)
                .orElseThrow(() -> new ErrorResponseException(HttpStatus.NOT_FOUND));
        
        return Optional.ofNullable(temp);
    }
```
OR 
Keep handle it with `.isEmpty `and then throw the error.
```
    public Optional<T> findById(long id) {
        var temp = someDao.findById(id);
        if (temp.isEmpty()){
            logger.error("ID not found");
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }

        return temp;
    }
``` 
is there pro/con between the two?

Reply:
The only difference is the log line in option 2. If you are throwing an exception you do not need to make the return type `Optional` you can make it `<T>` because there will never be an empty `Optional` (you will throw an exception when nothing is found).

Wether it's better or not there is no right answer... it just depends on what your application needs. Using `Optional<T>` as the return type for `findById` means that somewhere in your application if the record could not be found you do something in that case other than throw an error. If all you will always do is throw an error on an empty optional then you can just do that in the `findById` method.

A general note on api design is: for endpoints that get a single object / entity as return type we usually return `404 Not Found` when nothing is found. However for endpoints that return a list we return `200 Ok` but with an empty list as the response. This is something you apply to code too: never return `null` or `Optional` for methods that have a collection (array, list, set etc.) as return type. Just return an empty collection, it makes the code using that method much simpler to write since you can apply any operation (for, stream, size etc.) on an empty collection and you don't need to have any `if (list != null)` or `if (!listOptional.isEmpty())` checks

Reply:
Thanks, that makes sense. The way you phrased it is quite clear.